clinvar/src/main/scala/org/broadinstitute/monster/etl/clinvar/models/intermediate/VariationArchive.scala

```diff
@@ -250,57 +247,109 @@ object VariationArchive {
-              clinicalAssertionId = scv.id
+              id = s"${scvAccessionId}.${observationCounter.getAndIncrement()}",
+              clinicalAssertionId = scvAccessionId
             )
             rawObservation.extract("TraitSet").foreach { rawTraitSet =>
-              val traitSet =
-                SCVTraitSet.fromRawObservationSet(observation, rawTraitSet)
               val traitCounter = new AtomicInteger(0)
+              val currentScvTraitIds = new mutable.ArrayBuffer[String]()
               MsgTransformations.popAsArray(rawTraitSet, "Trait").foreach { rawTrait =>
                 val scvTrait = SCVTrait.fromRawTrait(
-                  traitSet,
+                  observation.id, // the setId is the same as the observation.id when extracting from observations
                   traitsWithoutContent,
                   relevantMappings,
                   traitCounter,
                   rawTrait
                 )
                 scvTraits.append(WithContent.attachContent(scvTrait, rawTrait))
+                currentScvTraitIds.append(scvTrait.id)
               }
+              val traitSet = SCVTraitSet.fromRawObservationSet(
+                observation,
+                rawTraitSet,
+                currentScvTraitIds.toArray
+              )
               scvTraitSets.append(WithContent.attachContent(traitSet, rawTraitSet))
             }
             observations.append(WithContent.attachContent(observation, rawObservation))
           }
 
           submitters.append(submitter)
           submissions.append(submission)
+
           // NOTE: It's important to attach content at the very end, to be sure everything
           // that can be modeled has already been popped out of the raw data.
+
+          val relevantTraitSetIds = scvTraitSets
+          // filter scvTraitSets down to the ones for the current scv
+            .filter(_.data.id == scvAccessionId)
+            .flatMap { traitSet =>
+              // need to use traitSet to find the right scv traits first
+              val scvTraitIds = scvTraits.filter { `trait` =>
+                traitSet.data.traitIds.contains(`trait`.data.id)
+              }.flatMap(_.data.traitId)
+              // for that set of scv traits, compare the traitIds (the clinvar ones) to the vcvTraits.
+              val targetTraits = scvTraitIds.intersect(vcvTraits.map(_.data.id))
+
+              vcvTraitSets
+              // filter vcvTraitSets down to the ones that have the same traits as the current scvTraitSet
+                .filter(_.data.traitIds sameElements targetTraits)
+                .map { filteredSet =>
+                  // get the IDs of the relevant vcvTraitSets
+                  filteredSet.data.id
+                }
+            }
+
+          val rcvId = rcvs.filter { rcv =>
+            // filter down to rcvs that contain the same traitSetIds (should be 1, no more no less)
+            relevantTraitSetIds.contains(rcv.data.traitSetId.getOrElse(None))
```

Comment:
Using `find` you could flip this condition to something like `rcv.traitSetId.isDefined && rcv.traitSetId == relevantTraitSetId`